File: Stage24/C/Problem11444.c
#include <stdio.h>

#define MOD 1000000007LL

typedef struct {
    long long a[2][2];
} Matrix;

Matrix multiply(Matrix x, Matrix y) {
    Matrix r;

    r.a[0][0] = (x.a[0][0]*y.a[0][0] + x.a[0][1]*y.a[1][0]) % MOD;
    r.a[0][1] = (x.a[0][0]*y.a[0][1] + x.a[0][1]*y.a[1][1]) % MOD;
    r.a[1][0] = (x.a[1][0]*y.a[0][0] + x.a[1][1]*y.a[1][0]) % MOD;
    r.a[1][1] = (x.a[1][0]*y.a[0][1] + x.a[1][1]*y.a[1][1]) % MOD;

    return r;
}

Matrix power(Matrix base, long long exp) {
    if (exp == 1) return base;

    Matrix half = power(base, exp / 2);
    Matrix ret = multiply(half, half);

    if (exp % 2 == 1)
        ret = multiply(ret, base);

    return ret;
}

int main(void) {
    long long n;
    scanf("%lld", &n);

    if (n == 0) {
        printf("0\n");
        return 0;
    }

    Matrix base = {{{1, 1}, {1, 0}}};
    Matrix result = power(base, n);

    printf("%lld\n", result.a[0][1]);
    return 0;
}

File: Stage24/Python/Problem11444.py
import sys
input = sys.stdin.readline

MOD = 1000000007

def multiply(a, b):
    return [
        [
            (a[0][0]*b[0][0] + a[0][1]*b[1][0]) % MOD,
            (a[0][0]*b[0][1] + a[0][1]*b[1][1]) % MOD
        ],
        [
            (a[1][0]*b[0][0] + a[1][1]*b[1][0]) % MOD,
            (a[1][0]*b[0][1] + a[1][1]*b[1][1]) % MOD
        ]
    ]

def power(matrix, exp):
    if exp == 1:
        return matrix

    half = power(matrix, exp // 2)
    ret = multiply(half, half)

    if exp % 2 == 1:
        ret = multiply(ret, matrix)

    return ret

def main():
    n = int(input())

    if n == 0:
        print(0)
        return

    base = [[1, 1], [1, 0]]
    result = power(base, n)

    print(result[0][1])

if __name__ == "__main__":
    main()